crates/iceberg/src/io/storage.rs

```diff
@@ -15,220 +15,296 @@
-    #[cfg(feature = "storage-azdls")]
-    Azdls {
-        /// Because Azdls accepts multiple possible schemes, we store the full
-        /// passed scheme here to later validate schemes passed via paths.
-        configured_scheme: AzureStorageScheme,
-        config: Arc<AzdlsConfig>,
-    },
-}
+use async_trait::async_trait;
+use bytes::Bytes;
 
-impl Storage {
-    /// Convert iceberg config to opendal config.
-    pub(crate) fn build(file_io_builder: FileIOBuilder) -> crate::Result<Self> {
-        let (scheme_str, props, extensions) = file_io_builder.into_parts();
-        let scheme = Self::parse_scheme(&scheme_str)?;
-
-        match scheme {
-            #[cfg(feature = "storage-memory")]
-            Scheme::Memory => Ok(Self::Memory(super::memory_config_build()?)),
-            #[cfg(feature = "storage-fs")]
-            Scheme::Fs => Ok(Self::LocalFs),
-            #[cfg(feature = "storage-s3")]
-            Scheme::S3 => Ok(Self::S3 {
-                configured_scheme: scheme_str,
-                config: super::s3_config_parse(props)?.into(),
-                customized_credential_load: extensions
-                    .get::<CustomAwsCredentialLoader>()
-                    .map(Arc::unwrap_or_clone),
-            }),
-            #[cfg(feature = "storage-gcs")]
-            Scheme::Gcs => Ok(Self::Gcs {
-                config: super::gcs_config_parse(props)?.into(),
-            }),
-            #[cfg(feature = "storage-oss")]
-            Scheme::Oss => Ok(Self::Oss {
-                config: super::oss_config_parse(props)?.into(),
-            }),
-            #[cfg(feature = "storage-azdls")]
-            Scheme::Azdls => {
-                let scheme = scheme_str.parse::<AzureStorageScheme>()?;
-                Ok(Self::Azdls {
-                    config: super::azdls_config_parse(props)?.into(),
-                    configured_scheme: scheme,
-                })
-            }
-            // Update doc on [`FileIO`] when adding new schemes.
-            _ => Err(Error::new(
-                ErrorKind::FeatureUnsupported,
-                format!("Constructing file io from scheme: {scheme} not supported now",),
-            )),
-        }
-    }
+use super::{Extensions, FileMetadata, FileRead, FileWrite, InputFile, OutputFile};
+use crate::Result;
+
+/// Trait for storage operations in Iceberg.
+///
+/// This trait defines the interface for all storage backends. Implementations
+/// provide access to different storage systems like S3, GCS, Azure, local filesystem, etc.
+///
+/// # Example
+///
+/// ```rust,ignore
+/// use iceberg::io::Storage;
+///
+/// async fn example(storage: Arc<dyn Storage>) -> Result<()> {
+///     // Check if file exists
+///     if storage.exists("s3://bucket/path/file.parquet").await? {
+///         // Read file
+///         let data = storage.read("s3://bucket/path/file.parquet").await?;
+///     }
+///     Ok(())
+/// }
+/// ```
+#[async_trait]
+pub trait Storage: Debug + Send + Sync {
```

Comment:
Following our discussion in last community sync, let's create a new crate called `iceberg-storage`, and put newly added things into that storage. I think with this approach we don't need to change existing code path, and when we reached consensus on the api, we could replace `FileIO` in core crate with `FileIO` in `iceberg-storage`.

Reply:
We discussed this offline and decided to move on by putting `Storage` trait in `iceberg` crate as a phase 1 implementation

We will need to move them to a new crate `iceberg-storage` after stabilizing it and moving on to phase 2